### ISSUE TITLE
nnet3-xvector-get-egs error

### ISSUE THREAD

**Issue by cathryn0131** [kaldi-asr/kaldi#2260]
I am trying to use my own data(the dim of feature is 1952 per frame, so I just extract 512 utterance to performance a subset training to verify the procedure ) to train an xvector, I used the xvector training script in sre16/v2, because I don't know the detailed format of the training data, I produced the corresponding preparation files by follow the sre16/v2 script, and now I met an error:

# nnet3-xvector-get-egs --compress=true --num-pdfs=6 exp/xvector_partial/egs/temp/ranges.1 "scp,s,cs:utils/filter_scp.pl exp/xvector_partial/egs/temp/ranges.1 data/xvector_partial/feats.scp |" ark:exp/xvector_partial/egs/egs_temp.1.ark
# Started at Wed Mar  7 10:15:14 CST 2018
#
nnet3-xvector-get-egs --compress=true --num-pdfs=6 exp/xvector_partial/egs/temp/ranges.1 'scp,s,cs:utils/filter_scp.pl exp/xvector_partial/egs/temp/ranges.1 data/xvector_partial/feats.scp |' ark:exp/xvector_partial/egs/egs_temp.1.ark
ASSERTION_FAILED (nnet3-xvector-get-egs[5.3]:SparseVector():sparse-matrix.cc:266) : 'pairs_.front().first >= 0 && pairs_.back().first < dim_'

[ Stack-Trace: ]
nnet3-xvector-get-egs() [0x5781ba]
kaldi::MessageLogger::HandleMessage(kaldi::LogMessageEnvelope const&, char const*)
kaldi::MessageLogger::~MessageLogger()
kaldi::KaldiAssertFailure_(char const*, char const*, int, char const*)
kaldi::SparseVector<float>::SparseVector(int, std::vector<std::pair<int, float>, std::allocator<std::pair<int, float> > > const&)
kaldi::SparseMatrix<float>::SparseMatrix(int, std::vector<std::vector<std::pair<int, float>, std::allocator<std::pair<int, float> > >, std::allocator<std::vector<std::pair<int, float>, std::allocator<std::pair<int, float> > > > > const&)
kaldi::nnet3::NnetIo::NnetIo(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int, std::vector<std::vector<std::pair<int, float>, std::allocator<std::pair<int, float> > >, std::allocator<std::vector<std::pair<int, float>, std::allocator<std::pair<int, float> > > > > const&, int)
nnet3-xvector-get-egs() [0x4b2376]
main
__libc_start_main
_start

# Accounting: time=0 threads=1
# Ended (code 134) at Wed Mar  7 10:15:14 CST 2018, elapsed time 0 seconds

Could anyone help check where is the root cause, thanks in advance.

**Comment by danpovey** [comment]
@david-ryan-snyder or anyone, could you please look into this?
even if the input data is bad, it shouldn't crash with an assertion error.

**Comment by david-ryan-snyder** [comment]
Yes... I'll have to add some more error handling.

@cathryn0131, my guess for now is that --num-pdfs=6 is less than the number of speakers in your ranges.1 file. Could you run something like `awk '{print $6}' exp/xvector_partial/egs/temp/ranges.1` | uniq and tell me what you see?

**Comment by cathryn0131** [comment]
@david-ryan-snyder 
awk '{print $6}' exp/xvector_partial/egs/temp/ranges.1 | uniq 
94
170
280
365
461
511
I extracted 512 utterances for 6 speakers, so there are only 6 speaker id.

**Comment by david-ryan-snyder** [comment]
Ok, these should be relabelled to be contiguous integers starting at 0, e.g., 0, 1, 2, etc. I think that will solve the problem (I should still add a check for this, though).

**Comment by cathryn0131** [comment]
@david-ryan-snyder 
I checked the script, the mapping from spk2utt to utt2int(in sid/nnet3/xvector/get_egs.sh line 112 to 116) to ranges.1(stage 2 in get_egs.sh,actually in allocate_egs.py).
actually it will calculate the number of utterances for each speaker, and transfer the number to speaker id in ranges.1
That is, 94 utterance for speaker1, 170 for speaker2, and so on. 
I don't know whether I can change the speaker id in ranges.1 to contiguous integers starting at 0, will the modification affect other latter operation?

    112   awk -v id=0 '{print $1, id++}' $data/spk2utt > $temp/spk2int
    113   utils/sym2int.pl -f 2 $temp/spk2int $data/utt2spk > $temp/utt2int
    114   utils/filter_scp.pl $temp/utt2num_frames.train $temp/utt2int > $temp/utt2int.train
    115   utils/filter_scp.pl $temp/utt2num_frames.valid $temp/utt2int > $temp/utt2int.valid
    116   utils/filter_scp.pl $temp/utt2num_frames.train_subset $temp/utt2int > $temp/utt2int.train_subset
//this is stage to to form ranges.1 file
    150 if [ $stage -le 2 ]; then
    151   echo "$0: Allocating training examples"
    152   $cmd $dir/log/allocate_examples_train.log \
    153     sid/nnet3/xvector/allocate_egs.py \
    154       --num-repeats=$num_repeats \
    155       --min-frames-per-chunk=$min_frames_per_chunk \
    156       --max-frames-per-chunk=$max_frames_per_chunk \
    157       --frames-per-iter=$frames_per_iter \
    158       --num-archives=$num_train_archives --num-jobs=$nj \
    159       --utt2len-filename=$dir/temp/utt2num_frames.train \
    160       --utt2int-filename=$dir/temp/utt2int.train --egs-dir=$dir  || exit 1

**Comment by david-ryan-snyder** [comment]
Unless you changed something, utt2int is a mapping from utterance to integer ID. It's not a count of the number of utterances per speaker.

**Comment by david-ryan-snyder** [comment]
The spk2int is normally a list of <string-id> to <int-id> pairs, where the int-ids are 0, 1, 2, 3, etc. This is what the script normally produces. Since it looks like you're getting something else, my guess is that something went wrong when you were preparing your data directory. It will probably be easier to just delete the data subset you were working on and start over (e.g., subset your original data directory again, rerun the egs extraction on it). Just make sure that you spk2int file (and later, the ranges files) have the right form.

**Comment by cathryn0131** [comment]
@david-ryan-snyder 
I followed your instruction, modified the speaker-id in utt2int to contiguous number 0-5, and skip 
utils/sym2int.pl -f 2 $temp/spk2int $data/utt2spk > $temp/utt2int
it works well and trained successfully.
I will check this sym2int script and update the cause later.
Thanks for your help.

**Comment by cathryn0131** [comment]
@danpovey 
hi, just one more question, when will MultiTask be working on NNET3, is there any schedule? I just see this on NNET1 now mainly for multilingual

**Comment by johnjosephmorgan** [comment]
I think you want to look at the babel_multilang recipe.
